From 36598ab6177f6a50b5c06b0cda2bf28b11579ff2 Mon Sep 17 00:00:00 2001
From: Thiago Macieira <thiago.macieira@intel.com>
Date: Wed, 4 Nov 2020 20:09:41 -0800
Subject: [PATCH 044/171] QMap: don't tell everyone QMapNode has no friends
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

 qmap.h:107:8: warning: ‘struct QMapNode<Key, T>’ only defines private constructors and has no friends [-Wctor-dtor-privacy]

QMapNode is created by QMapNodeBase::createNode, which is given the full
size and alignment. But nowhere do we actually call the constructor for
this type, which means, strictly speaking, the lifetime for QMapNode
never begins. That is, we have UB here.

By adding the friend declaration, we neatly shut up the compiler since
it can't know that the called function can't call the constructor.
Though an even more aggressive compiler could make this assumption for
types in unnamed namespaces: if it can't see the body of createNode(),
it could correctly deduce that the constructor is never called and
therefore the lifetime has never begun.

This commit does not fix the actual UB. It just silences the warning.

Doesn't apply to Qt 6 since there QMap was rewritten on top of std::map.

Task-number: QTBUG-88185
Change-Id: I7b9b97ae9b32412abdc6fffd164481295e85db19
Reviewed-by: Giuseppe D'Angelo <giuseppe.dangelo@kdab.com>
---
 src/corelib/tools/qmap.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/corelib/tools/qmap.h b/src/corelib/tools/qmap.h
index d72f5e574b..676d9a1fec 100644
--- a/src/corelib/tools/qmap.h
+++ b/src/corelib/tools/qmap.h
@@ -141,6 +141,7 @@ private:
 
     QMapNode() = delete;
     Q_DISABLE_COPY(QMapNode)
+    friend struct QMapNodeBase;
 };
 
 template <class Key, class T>
-- 
2.31.1

